### PR TITLE
Version 2.0.0

### PR DESCRIPTION
**Pull request by hirshbergshachar**

### CODE REVIEW

File: add-on/TA-Demisto/bin/demisto_setup.py [hunk 1]
```diff
@@ -91,80 +89,90 @@ def handleEdit(self, confInfo):
                 url += ":" + self.callerArgs.data['PORT'][0]
 
             '''
-                Create Test Incident to demisto to verify if the configuration entered are correct
+                Create Test Incident to demisto to verify if the configuration is correct
                 Store configuration only if create incident was successful.
             '''
             url += "/incidenttype"
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_setup.py]: Instead of directly going to `/incidenttype` that have authorization and can fail for multiple reasons, I suggest we will make the following calls so we will give a proper message to the user what have failed.
1. `/proxyMode` - no auth, so if we fail here it is just about network / certificate.
2. `/user` - will let us check auth and that this user have the required permissions.

For each case we should send a message to the user with the exact reason, ad if we have a suggestion include it.

**Comment by orius123** [comment]
@hirshbergshachar is it ready?

File: add-on/TA-Demisto/bin/demisto_setup.py (hunk 1 above)
Comment (hirshbergshachar) [reply add-on/TA-Demisto/bin/demisto_setup.py]: @orius123 added.

**Comment by hirshbergshachar** [comment]
@orius123 it is ready.

**Comment by orius123** [comment]
GR8 JOB!  💪 

I'll review later this week

File: add-on/TA-Demisto/README.md [hunk 2]
```diff
@@ -21,16 +19,57 @@ Supporting Add-on for Demisto. This application allows a user to create incident
 * This app can be installed through UI using "Manage Apps" or from the command line using the following command:
 $SPLUNK_HOME/bin/splunk install app $PATH_TO_SPL/TA-Demisto.spl/
 
+# HTTPS Certificate Validation
+
+The add-on uses secure communication with Demisto.
+
+On few occasions it would be necessary to manually set a certificate for proper verification.
```
Comment (orius123) [inline add-on/TA-Demisto/README.md:26]: Should mention that users that had previous versions of the app and disabled certificate will have to do it again.

File: add-on/TA-Demisto/README.md [hunk 3]
```diff
@@ -21,16 +19,57 @@ Supporting Add-on for Demisto. This application allows a user to create incident
...
+
+We recommend to use certificates, and only disabling certificate verification in development
+or test environments only. Never disable certificate verification for a production system.
+
+* User can directly extract the app's SPL file into $SPLUNK_HOME/etc/apps/ folder in order to install the app.
 
 # Application Setup
-* The user must complete the setup of the application. In order to create incident into Demisto, a user needs to provide following four parameters:
+* The user must complete the setup of the application. In order to create incident into Demisto, a user needs to provide the following parameters:
     1)Demisto URL: This is mandatory parameter Url /IP address of the Demisto
```
Comment (orius123) [inline add-on/TA-Demisto/README.md]: Add a space after the number

File: add-on/TA-Demisto/README.md [hunk 4]
```diff
@@ -3,12 +3,10 @@
 
 Supporting Add-on for Demisto. This application allows a user to create incident into Demisto from Splunk using custom alert action.
 
-
 # REQUIREMENTS
 
 * Splunk version 6.3 >=
-* This application should be installed on Search Head.
-
+* This application should be installed on Search Head or Universal forwarder, Indexer, and Search Head if applicable.
```
Comment (orius123) [inline add-on/TA-Demisto/README.md:9]: Why search head twice?

File: add-on/TA-Demisto/README.md [hunk 5]
```diff
@@ -21,16 +19,57 @@ Supporting Add-on for Demisto. This application allows a user to create incident
...
+    2.1) Extract TA-Demisto tgz
+    2.2) Create "local" directory under TA-Demisto
+    2.3) Place the certificate in that folder
+    In the end the certificate should be in the following path:
+    your-local-path/TA-Demisto/local/cert_bundle.pem
+
+Splunk Cloud users should perform option 2 and send the app installer to Splunk support for installation.
 
-* User can directly extract SPL file  into $SPLUNK_HOME/etc/apps/ folder.
+Another option which applies ONLY for on-prem installations and is cannot be used for Splunk Cloud:
```
Comment (orius123) [inline add-on/TA-Demisto/README.md]: cannot without the 'is'

File: add-on/TA-Demisto/README.md [hunk 6]
```diff
@@ -21,16 +19,57 @@ Supporting Add-on for Demisto. This application allows a user to create incident
...
     1)Demisto URL: This is mandatory parameter Url /IP address of the Demisto
     2)Demisto Port: This is an optional parameter. The user must define it if running Demisto on any other port than the default (443).
-    3)Allow Self Signed Certificate: User should select this if using Self Signed certificate
-    4) Authentication key: This is a mandatory parameter. This parameter is used for authorization with Demsito. In order to generate this parameter, a user should log in to Demisto and then click on Settings --> Integration --> API Keys.
+    3)Authentication key: This is a mandatory parameter. This parameter is used for authorization with Demisto. In order to generate this parameter,
+      a user should log in to Demisto and then click on Settings --> Integration --> API Keys.
+    4)Location to certificate: Enter the full path to the SSL Certificate in the Splunk server to if you are using Self Signed/Internal CA signed certificate
+    5)HTTPS Proxy Address: This is an optional parameter. Define this if you have HTTPS proxy that should be used. Should be in the format of https://ip:port
```
Comment (orius123) [inline add-on/TA-Demisto/README.md]: There is a standard for proxy with creds. why don't we use it? 
`user:password@proxyIP:proxyPort`

File: add-on/TA-Demisto/README.md [hunk 7]
```diff
@@ -41,9 +80,24 @@ $SPLUNK_HOME/bin/splunk install app $PATH_TO_SPL/TA-Demisto.spl/
...
+  Afterwards, when you'll create a new alert, choose 'Create Demisto Incident' and you'll have an option 'Demisto Server' where you can choose the server to send the alert to.
+
+  If you want to send alert to all of your Demisto servers, simply click the 'send alert to all of the servers' checkbox. No need to select a server in this scenario.
+
 #Troubleshooting
 * Environment variable SPLUNK_HOME must be set
-* To troubleshoot Demisto application, check $SPLUNK_HOME/var/log/splunk/demisto.log file.
+* Check the following logs to troubleshoot Demisto's application:
+    1) $SPLUNK_HOME/var/log/demisto/demisto.log file
+    2) $SPLUNK_HOME/var/log/splunk/demisto_modalert.log
```
Comment (orius123) [inline add-on/TA-Demisto/README.md]: `splunkd.log` is way more important than `...modalert.log`

File: add-on/TA-Demisto/bin/demisto_send_alert.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+            for entry in conf_dic.get('entry', []):
+                val = entry.get('content', {})
+                if val:
+                    config = val
+        if '' in config:
+            config.pop('')
+        if 'config' in config:
+            config.pop('config')
+
+        demisto_servers = config.get('DEMISTOURL', '').split(',')
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_send_alert.py]: trim also

File: add-on/TA-Demisto/bin/demisto_config.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            for entry in conf_dic.get('entry', []):
+                val = entry.get('content', {})
+                if val:
+                    config = val
+        if '' in config:
+            config.pop('')
+        if 'config' in config:
+            config.pop('config')
+
+        servers = config.get('DEMISTOURL', '').split(',')
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_config.py]: should trim here as well?

File: add-on/TA-Demisto/bin/demisto_config.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        get_args = {
+            'output_mode': 'json'
+        }
+
+        success, content = splunk.rest.simpleRequest(CONFIG_ENDPOINT, self.sessionKey, method='GET',
+                                                     getargs=get_args)
+
+        conf_dic = json.loads(content)
+        config = {}
+        if success and conf_dic:
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_config.py:69]: Don't we want to log if we haven't found anything?

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        self.logger = logger
+
+    def create_incident(self, authkey, data, verify_req, search_query="", search_url="", ssl_cert_loc="",
+                        result=None, search_name=None, proxies=None, url=None):
+        """
+            This method is used to create incident in Demisto. It takes four arguments and all are mandatory:
+            @url: Demisto URL, its mandatory parameter.
+            @authkey: Requires parameter, used for authentication.
+            @data: Incident information,
+            @verify: Indicates if self signed certificates are allowed or not.
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: arg name is `verify_req`

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        """
+
+        incident = self.create_incident_dictionary(data, search_query, search_url, result, search_name)
+
+        s = requests.session()
+
+        # url is either the selected Demisto server, or sending to all of the available servers
+        if not url:
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: when will this be true?

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        self.logger.debug("JSON data for the Incident=" + json.dumps(incident))
+
+        req = Request('POST', url + "/incident/splunkapp", data=json.dumps(incident))
+        prepped = s.prepare_request(req)
+
+        prepped.headers['Authorization'] = authkey
+        prepped.headers['Content-type'] = "application/json"
+        prepped.headers['Accept'] = "application/json"
+
+        if ssl_cert_loc:
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: If user will disable the certificate check but will save path in config you will still try to use it, this is confusing.
If it's disabled should always be disabled

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                                             time.localtime(int(float(data["occured"])))) + timezone
+
+        # Always pass True for create investigation
+        incident["createInvestigation"] = True
+
+        incident["name"] = data.get('incident_name', '')
+
+        incident["type"] = data.get('type', '')
+
+        ignore_labels = data.get('ignore_labels', '').lower().split(",")
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: trim

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+        if search_name:
+            data_dir = {'type': 'search_name', 'value': search_name}
+            labels.append(data_dir)
+
+        self.logger.debug("Label::::" + str(result.keys()))
+        self.logger.debug("Ignore Label::::" + str(ignore_labels))
+
+        if data.get("labels"):
+            str_data = data["labels"].split(",")
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: trim

File: add-on/TA-Demisto/bin/demisto_incident.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        else:
+            for key in result.keys():
+                if key.lower() not in ignore_labels and not key.startswith("__"):
+                    data_dir = {"type": key, "value": result[key]}
+                    labels.append(data_dir)
+
+        incident["labels"] = labels
+
+        if "custom_field" in data:
+            str_data = data["custom_field"].split(",")
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_incident.py]: again. trim.
Users may time add spaces after commas.
you should validate you have strings before trimming.
suggest to create a def for split and trim

File: add-on/TA-Demisto/bin/demisto_send_alert.py [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+                "Demisto Incident creation in create_demisto_incident function failed. exception=" + str(ex),
+                sourcetype="demistoResponse")
+
+    def get_password_for_server(self, save_name):
+        r = splunk.rest.simpleRequest(SPLUNK_PASSWORD_ENDPOINT, self.session_key, method='GET', getargs={
+            'output_mode': 'json', 'search': save_name})
+
+        password = ""
+
+        if 200 <= int(r[0]["status"]) < 300:
```
Comment (orius123) [inline add-on/TA-Demisto/bin/demisto_send_alert.py]: many things can go wrong here. wrap in expect

File: add-on/TA-Demisto/default/setup.xml [hunk 18]
```diff
@@ -1,19 +1,19 @@
...
+	{
 		var app = Splunk.util.getCurrentApp();
 		var css = "/static/app/" + app + "/setup.css";
 		addStylesheet(css);
 
-		$(window).keydown(function(event){			
+		$(window).keydown(function(event){
 			if(event.keyCode == 13) {
-				if(validateTextBox("DEMISTOURL") && validateNumber("PORT") &&  validateTextBox("AUTHKEY") && validateHostName("DEMISTOURL"))
+				if(validateTextBox("AUTHKEY"))
```
Comment (orius123) [inline add-on/TA-Demisto/default/setup.xml]: why did you remove port and host validations?

**Review by orius123 — COMMENTED** [review]
Hi, 
This is my 2 cents.

**Comment by orius123** [comment]
Please have alex review the README as well

File: add-on/TA-Demisto/README.md (hunk 4 above)
Comment (hirshbergshachar) [reply add-on/TA-Demisto/README.md:9]: @orius123 it's two different installation constellation. Thought it's better to specify instead of saving in words

**Comment by meirwah** [comment]
@hirshbergshachar plz note the review

File: add-on/TA-Demisto/bin/demisto_config.py (hunk 10 above)
Comment (hirshbergshachar) [reply add-on/TA-Demisto/bin/demisto_config.py:69]: This is validated in another part of the code.

File: add-on/TA-Demisto/README.md (hunk 6 above)
Comment (hirshbergshachar) [reply add-on/TA-Demisto/README.md]: @orius123 we can change that part of the code in the next version.

**Comment by orius123** [comment]
👏 Hoorray!
@hirshbergshachar don't forget to close the issue for multiple servers
